node/core/approval-voting/src/lib.rs

```diff
@@ -539,6 +533,95 @@ struct ApprovalStatus {
 	block_tick: Tick,
 }
 
+enum ApprovalOutcome {
+	Approved,
+	Failed(InvalidCandidate),
+}
+
+struct ApprovalState {
+	validator_index: ValidatorIndex,
+	candidate_hash: CandidateHash,
+	candidate_index: CandidateIndex,
+	status: ApprovalOutcome,
+}
+
+impl ApprovalState {
+	fn approved(
+		validator_index: ValidatorIndex,
+		candidate_hash: CandidateHash,
+		candidate_index: CandidateIndex
+	) -> Self {
+		Self {
+			validator_index,
+			candidate_hash,
+			candidate_index,
+			status: ApprovalOutcome::Approved
+		}
+	}
+	fn failed(
+		validator_index: ValidatorIndex,
+		candidate_hash: CandidateHash,
+		candidate_index: CandidateIndex,
+		reason: InvalidCandidate
+	) -> Self {
+		Self {
+			validator_index,
+			candidate_hash,
+			candidate_index,
+			status: ApprovalOutcome::Failed(reason)
+		}
+	}
+}
+
+type BoundedBackgroundApprovalTask = Timeout<RemoteHandle<ApprovalState>>;
+
+struct CurrentlyCheckingSet {
+	approvals_cache: lru::LruCache<CandidateHash, ApprovalOutcome>,
```

Comment:
`approvals_cache` doesn't look to be used here, so it should be moved out of the struct. I think it's correct not to use it here FWIW